src/server/services/procedures/cloud-variables/cloud-variables.js

```diff
@@ -396,21 +391,115 @@ CloudVariables.setUserVariable = function(name, value) {
-    return userVars.deleteOne({name: name, owner: username})
-        .then(() => 'OK');
+    await userVars.deleteOne({name: name, owner: username});
+    delete (userListeners[username] || {})[name];
+};
+
+/**
+ * Equivalent to calling :func:`CloudVariables.lockVariable` followed by :func:`CloudVariables.getVariable`.
+ * @param {String} name Variable name
+ * @param {String=} password Password (if password-protected)
+ */
+CloudVariables.lockAndGetVariable = async function(name, password) {
+    await this.lockVariable(name, password);
+    return await this.getVariable(name, password);
+};
+
+/**
+ * Equivalent to calling :func:`CloudVariables.setVariable` followed by :func:`CloudVariables.unlockVariable`.
+ * @param {String} name Variable name
+ * @param {Any} value Value to store in variable
+ * @param {String=} password Password (if password-protected)
+ */
+CloudVariables.setAndUnlockVariable = async function(name, value, password) {
+    await this.setVariable(name, value, password);
+    await this.unlockVariable(name, password);
+};
+
+/**
+ * Sets the variable to the given value and returns the previous value.
+ * @param {String} name Variable name
+ * @param {Any} value Value to store in variable
+ * @param {String=} password Password (if password-protected)
+ */
+CloudVariables.getAndSetVariable = async function(name, value, password) {
+    const res = await this.getVariable(name, password);
+    await this.setVariable(name, value, password);
+    return res;
+};
+
+/**
+ * Sets the variable to the given value and returns the previous value.
+ * @param {String} name Variable name
+ * @param {Any} value Value to store in variable
+ */
+CloudVariables.getAndSetUserVariable = async function(name, value) {
+    const res = await this.getUserVariable(name);
+    await this.setUserVariable(name, value);
+    return res;
+};
+
+CloudVariables._getListenBucket = function (name) {
+    let bucket = globalListeners[name];
+    if (!bucket) bucket = globalListeners[name] = {};
+    return bucket;
+};
+CloudVariables._getUserListenBucket = function (name) {
+    const user = this.caller.username;
+    let userBucket = userListeners[user];
+    if (!userBucket) userBucket = userListeners[user] = {};
+
+    let bucket = userBucket[name];
+    if (!bucket) bucket = userBucket[name] = {};
+    return bucket;
+};
+
+/**
+ * Registers your client to receive messages each time the variable value is updated.
+ * ``name`` and ``password`` denote the variable to listen to.
+ * ``msgType`` is the name of the message that will be sent each time it is updated.
+ * 
+ * The variable must already exist prior to calling this RPC.
+ * Update events will cease when the variable is deleted.
+ * 
+ * **Message Fields**
+ * 
+ * - ``name`` - the name of the variable that was updated
+ * - ``value`` - the new value of the variable
+ * 
+ * @param {String} name Variable name
+ * @param {Any} msgType Message type to send each time the variable is updated
```

Comment:
Shouldn't this be `String`? It could probably be made an optional value if we can come up with a reasonable default.

Reply:
Yeah, that should definitely be `String`. We could probably make a default like `'${var} changed'`, which would work for any number of vars, but idk if that'd be intuitive (without looking at docs)

Reply:
The default name of the message type doesn't need to be dynamic, does it? It looks like the name (and value) of the changed variable is included in the message itself.

Reply:
Oh, true... I was thinking from a usage perspective I'd (usually) prefer them be different message types, but a default single one would still work